bindings/node/test/autoEncrypter.test.js

```diff
@@ -0,0 +1,153 @@
+  const ejson = EJSON.parse(fs.readFileSync(path, 'utf8'));
+  return BSON.serialize(ejson);
+}
+
+function readHttpResponse(path) {
+  let data = fs.readFileSync(path, 'utf8').toString();
+  data = data.split('\n').join('\r\n');
+  return Buffer.from(data, 'utf8');
+}
+
+const TEST_COMMAND = JSON.parse(fs.readFileSync(`${__dirname}/data/cmd.json`));
+const MOCK_COLLINFO_RESPONSE = readExtendedJsonToBuffer(`${__dirname}/data/collection-info.json`);
+const MOCK_MONGOCRYPTD_RESPONSE = readExtendedJsonToBuffer(
+  `${__dirname}/data/mongocryptd-reply.json`
+);
+const MOCK_KEYDOCUMENT_RESPONSE = readExtendedJsonToBuffer(`${__dirname}/data/key-document.json`);
+const MOCK_KMS_DECRYPT_REPLY = readHttpResponse(`${__dirname}/data/kms-decrypt-reply.txt`);
+
+class MockClient {
+  constructor() {
+    this.topology = {
+      bson: BSON
+    };
+  }
+}
+
+const AutoEncrypter = require('../lib/autoEncrypter').AutoEncrypter;
+describe('AutoEncrypter', function() {
+  let ENABLE_LOG_TEST = false;
+  let sandbox = sinon.createSandbox();
+  before(() => {
+    sandbox.stub(StateMachine.prototype, 'kmsRequest').callsFake(request => {
+      request.addResponse(MOCK_KMS_DECRYPT_REPLY);
+      return Promise.resolve();
+    });
+
+    sandbox
+      .stub(StateMachine.prototype, 'fetchCollectionInfo')
+      .callsFake((client, ns, filter, callback) => {
+        callback(null, MOCK_COLLINFO_RESPONSE);
+      });
+
+    sandbox
+      .stub(StateMachine.prototype, 'markCommand')
+      .callsFake((client, ns, command, callback) => {
+        if (ENABLE_LOG_TEST) {
+          const response = BSON.deserialize(MOCK_MONGOCRYPTD_RESPONSE);
+          response.schemaRequiresEncryption = false;
+
+          ENABLE_LOG_TEST = false; // disable test after run
+          callback(null, BSON.serialize(response));
+          return;
+        }
+
+        callback(null, MOCK_MONGOCRYPTD_RESPONSE);
+      });
+
+    sandbox.stub(StateMachine.prototype, 'fetchKeys').callsFake((client, ns, filter, callback) => {
+      // mock data is already seriaized, our action deals with the result of a cursor
+      const deserializedKey = BSON.deserialize(MOCK_KEYDOCUMENT_RESPONSE);
+      callback(null, [deserializedKey]);
+    });
+  });
+
+  after(() => {
+    sandbox.restore();
+  });
+
+  describe('state machine', function() {
+    it('should decrypt mock data', function(done) {
+      const input = readExtendedJsonToBuffer(`${__dirname}/data/encrypted-document.json`);
+      const client = new MockClient();
+      const mc = new AutoEncrypter(client, {
+        keyVaultNamespace: 'admin.datakeys',
+        logger: () => {},
+        kmsProviders: {
+          aws: { accessKeyId: 'example', secretAccessKey: 'example' },
+          local: { key: Buffer.alloc(96) }
+        }
+      });
+
+      mc.decrypt(input, (err, decrypted) => {
+        if (err) return done(err);
+        expect(decrypted).to.eql({ filter: { find: 'test', ssn: '457-55-5462' } });
```

Comment:
This is just because the `encrypted-document.json` file we are using has an encrypted BSON document that is a filter for a find operation. Just a coincidence, could have been `{ llamas: 'are cool' }`